mlflow/pyfunc/__init__.py

```diff
@@ -266,7 +270,117 @@ def _load_model_env(path):
+    directly. Instead, instances of this class are constructed and returned from
+    py:func:`mlflow.pyfunc.load_model`.
+
+    ``model_impl`` can be any Python object that implements the `Pyfunc interface
+    <https://mlflow.org/docs/latest/python_api/mlflow.pyfunc.html#pyfunc-inference-api>`_, and is
+    by invoking the model's ``loader_module``.
+
+    ``model_meta`` contains model metadata loaded from the MLmodel file.
+    """
+
+    def __init__(self, model_meta: Model, model_impl: Any):
+        if not hasattr(model_impl, "predict"):
+            raise MlflowException("Model implementation is missing required predict method.")
+        self._model_meta = model_meta
+        self._model_impl = model_impl
+
+    def predict(self, data: pandas.DataFrame) -> PyFuncOutput:
+        """
+        Generate model predictions.
+        :param data: Model input as pandas.DataFrame.
+        :return: Model predictions as one of pandas.DataFrame, pandas.Series, numpy.ndarray or list.
+        """
+        input_schema = self._model_meta.get_input_schema()
+
+        if input_schema is not None:
+            def convert_type(name, values: pandas.Series, t: DataType):
+                if values.dtype == np.object and t not in (DataType.binary, DataType.string):
+                    values = values.infer_objects()
+
+                if values.dtype in (t.to_pandas(), t.to_numpy()):
+                    return values
+
+                if t == DataType.binary and values.dtype.kind == t.binary.to_numpy().kind:
+                    #  NB: bytes in numpy have variable itemsize and would fail the upcast check.
+                    return values
+
+                if t == DataType.string and values.dtype == np.object:
+                    #  NB: strings are by default parsed and inferred as objects, but it is
+                    # recommended to use StringDtype extension type if available. See
+                    #
+                    # `https://pandas.pydata.org/pandas-docs/stable/user_guide/text.html`
+                    #
+                    # for more detail.
+                    try:
+                        return values.astype(t.to_pandas(), errors="raise")
+                    except ValueError:
+                        raise MlflowException(
+                            "Failed to convert column {0} from type {1} to {2}.".format(
+                              name, values.dtype, t)
+                        )
+
+                numpy_type = t.to_numpy()
+                is_compatible_type = values.dtype.kind == numpy_type.kind
+                is_upcast = values.dtype.itemsize <= numpy_type.itemsize
+                if is_compatible_type and is_upcast:
+                    # conversions between compatible types and ints -> floats are allowed.
+                    return values.astype(numpy_type, errors="raise")
+                else:
+                    # NB: conversion between incompatible types (e.g. floats -> ints or
```

Comment:
or double -> float, may be good to point out

Reply:
kk makes sense